### ISSUE TITLE
Define Protocol API and State Streamer & Registration service

### ISSUE THREAD

**Issue by franklywatson** [onflow/flow-go#1989]
With the upcoming refactoring of Access Node and changing of how we incorporate DPS into the Observer we need to make sure that new key interfaces between components are specified up front. Using this epic to track this effort

**Comment by AliSMAmin** [comment]
1.  In terms of the State Streamer inputs, am I correct that it should only:

- Register itself as a new streamer to the registration API;
- Register block consumers, and;
- Register execution data consumers? 
- The registration API takes care of consuming finalized block and execution data, then feeding it with a pub/sub connection.

2. What kind of query parameters are we envisioning the State Streamer of being capable of handling?
3. Is it downloading the entire chain from genesis or just from the latest spork, then queries are funneled through Flow API/DPS API in target architecture? 
4. Should queries look similar to Protocol API for example? 
5. Or is the idea that the Flow API will determine what queries should ultimately go to Protocol API versus DPS API, and that things persisted in the Protocol DB or transactions go to Protocol API whereas queries for things like a specific transaction at particular a block height go to the DPS API and then ultimately to the State Streamer if not persisted in IndexDB?
6. Are you guys also looking to use Redoc for the documentation? I would imagine end-users would appreciate something on docs.onflow for service registration.

In terms of errors, I think that's self-explanatory.

**Comment by AliSMAmin** [comment]
@smnzhu Lokesh and I were wondering why in [state_streamer/state_streamer.go](https://github.com/onflow/flow-go/pull/2094/files) the subscriber is referred to as a Consumer `RegisterBlockConsumer(FinalizedBlockConsumer)` whereas in [protobuf/flow/state_streamer/state_streamer.proto](https://github.com/onflow/flow/pull/851/files) it is referred to as a subscription `GetBlocks(BlockSubscription) returns (stream entities.Block)` . Is it better to standardize the methods or just implement as a consumer with a subscription instead of a subscriber?

`RegisterBlockSubscriber(FinalizedBlockSubscriber)` instead of `RegisterBlockConsumer(FinalizedBlockConsumer)`
 and `RegisterExecutionDataSubscriber(ExecutionDataSubscriber)` instead of `RegisterExecutionBlockConsumer(ExecutionDataConsumer)` .

**Comment by peterargue** [comment]
After discussing this a bit more, we have a slightly altered proposal:

Instead of a state streamer api that contains just the pub/sub endpoints for the protocol and execution state, we’ll have 2 apis:
1. Protocol State API which is the protocol api we’ve described, plus the blocks pub/sub endpoint
* the subscribe endpoint will take a single argument that specifies the start block, which can be either a height or sealed/final
2. Execution State API which has an endpoint to request execution data for specific blocks, plus the pub/sub endpoint
* the subscribe endpoint will take a single argument that specifies the start block, which can be either a height or sealed/final

The intention for this change is have a more unified interface for clients, so they only need to interact with a single API to access a single data type.

**Comment by peterargue** [comment]
I think my last comment addresses 1&2.

For 3-5, are you asking generally, or are those questions specific to the streaming apis?

3. A pub/sub client should be able to access all blocks available on the AN/observer. Currently, that's limited to the blocks within a spork.
4. I'm not sure what you mean. Can you rephrase the question?
5. Yes, the idea is that the Flow API Service, will work as a kind of API Gateway and decide if the request should go to the protocol API, DPS or some other future API. DPS would get its data from the pub/sub APIs.

6. Having some autogenerated documentation would be great. We should discuss this more.

**Comment by AliSMAmin** [comment]
Correct me if wrong, but cursorily it looks like issue #1989 can be broken up into https://github.com/onflow/flow-go/issues/2318 and https://github.com/onflow/flow-go/issues/2319 , just adding pub/sub endpoints. I will sit down and explain my reasoning in-depth later.

**Comment by peterargue** [comment]
Yea I think that's right. I'm not totally clear on what the proposal is in #2319, but we'll need 2 new APIs:
* Protocol API which provides access to the protocol state, including a streaming endpoint for newly finalized blocks
* Execution State API which provides access to the execution state, including a streaming endpoint for newly downloaded execution data

**Comment by AliSMAmin** [comment]
That clears it up for us. Access API will essentially be copied into a new service, then refactored into Protocol API. However, instead of a separate State Streamer, it will have those streaming endpoints only for protocol data. 

Instead of a State Streamer handling protocol and execution data, we will build out an Execution State API - which also has steaming endpoints only for execution data.
